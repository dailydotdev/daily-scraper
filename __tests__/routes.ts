import { FastifyInstance } from 'fastify';
import * as request from 'supertest';
import { setupStaticServer } from './helpers';
import appFunc from '../src';

let fileServer: FastifyInstance;
let app: FastifyInstance;

beforeAll(async () => {
  app = appFunc();
  return app.ready();
});

afterEach(() => fileServer.close());

afterAll(() => app.close());

it('should return bad request when url is not provided', async () => {
  fileServer = await setupStaticServer('rss.xml', 'source.html');
  await request(app.server).get('/scrape/source').expect(400);
});

it('should scrape the website for source information', async () => {
  fileServer = await setupStaticServer('rss.xml', 'source.html');
  const res = await request(app.server)
    .get('/scrape/source')
    .query({ url: 'http://localhost:6789/' })
    .expect(200);
  expect(res.body).toMatchSnapshot();
});

it('should scrape even if not all information is available', async () => {
  fileServer = await setupStaticServer('rss.xml', 'partialSource.html');
  const res = await request(app.server)
    .get('/scrape/source')
    .query({ url: 'http://localhost:6789/' })
    .expect(200);
  expect(res.body).toMatchSnapshot();
});

it('should scrape rss feed and crawl the website', async () => {
  fileServer = await setupStaticServer('rss.xml', 'multipleSources.html');
  const res = await request(app.server)
    .get('/scrape/source')
    .query({ url: 'http://localhost:6789/rss.xml' })
    .expect(200);
  expect(res.body).toMatchSnapshot();
});

it('should scrape rss feed even if website is not available', async () => {
  fileServer = await setupStaticServer('rss.xml');
  const res = await request(app.server)
    .get('/scrape/source')
    .query({ url: 'http://localhost:6789/rss.xml' })
    .expect(200);
  expect(res.body).toMatchSnapshot();
});

it('should return 200 even if website is not available', async () => {
  fileServer = await setupStaticServer();
  const res = await request(app.server)
    .get('/scrape/source')
    .query({ url: 'http://localhost:6789/notfound' })
    .expect(200);
  expect(res.body).toMatchSnapshot();
});

it('should return the number of voters', async () => {
  fileServer = await setupStaticServer();
  const res = await request(app.server)
    .get('/scrape/mediumVoters')
    .query({ url: 'http://localhost:6789/mediumVoters.html' })
    .expect(200);
  expect(res.body).toEqual({ voters: 16 });
});

it('should return failed response when cannot find the number of voters', async () => {
  fileServer = await setupStaticServer();
  const res = await request(app.server)
    .get('/scrape/mediumVoters')
    .query({ url: 'http://localhost:6789/noMediumVoters.html' })
    .expect(200);
  expect(res.body).toEqual({ failed: true });
});

it('should indicate that the medium article does not exist', async () => {
  fileServer = await setupStaticServer();
  const res = await request(app.server)
    .get('/scrape/mediumVoters')
    .query({ url: 'http://localhost:6789/notfound.html' })
    .expect(200);
  expect(res.body).toEqual({ failed: true, error: '4xx' });
});
