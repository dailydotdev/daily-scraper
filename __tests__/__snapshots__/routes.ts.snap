// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return 200 even if website is not available 1`] = `
Object {
  "type": "unavailable",
}
`;

exports[`should scrape even if not all information is available 1`] = `
Object {
  "logo": "https://daily.dev/favicon-logo-192x192.png",
  "name": "Daily",
  "rss": Array [],
  "type": "website",
  "website": "http://localhost:6789/",
}
`;

exports[`should scrape rss feed and crawl the website 1`] = `
Object {
  "logo": "https://daily.dev/favicon-logo-192x192.png",
  "name": "Daily",
  "rss": Array [
    Object {
      "title": "RSS",
      "url": "http://localhost:6789/rss.xml",
    },
  ],
  "type": "website",
  "website": "http://localhost:6789/",
}
`;

exports[`should scrape rss feed even if website is not available 1`] = `
Object {
  "logo": "https://res.cloudinary.com/daily-now/image/upload/logos/placeholder.jpg",
  "name": null,
  "rss": Array [
    Object {
      "title": "RSS",
      "url": "http://localhost:6789/rss.xml",
    },
  ],
  "type": "website",
  "website": "http://localhost:6789/",
}
`;

exports[`should scrape the website for source information 1`] = `
Object {
  "logo": "https://daily.dev/favicon-logo-192x192.png",
  "name": "Daily",
  "rss": Array [
    Object {
      "title": "RSS",
      "url": "http://localhost:6789/rss.xml",
    },
  ],
  "type": "website",
  "website": "http://localhost:6789/",
}
`;
